Improve ArrayRef

smessmer opened the pull request:
Summary: Mostly making some stuff in ArrayRef constexpr to give it better perf.

Differential Revision: D8926785

colesbury commented:
In what sort of use cases did you see a performance improvement?

smessmer commented:
Sorry, I probably formulated that a bit misleadingly. It doesn't improve perf for existing use cases, but it allows ArrayRef to be used in constexpr environments in future (for example for c10).

Constexpr per se doesn't improve perf. Some expressions can be evaluated at compile time and some cannot. For expressions that can be evaluated at compile time, many compilers will do so, even if you don't add the "constexpr" keyword. What the "constexpr" keyword does is it asserts that the expression can be evaluated at compile time. So it makes sure that the stuff you want to be run at compile time actually can be done so.

ezyang commented on aten/src/ATen/ArrayRef.h:
```diff
@@ -23,170 +24,184 @@
+
+  /// Construct an empty ArrayRef.
+  // TODO Make this explicit
+  constexpr ArrayRef() : Data(nullptr), Length(0) {}
```
C++ noob question: why does explicit matter here? There's no implicit conversion to have created?

ezyang replied:
Also, why did you delete the `/* implicit */` comments? They seem like useful reminders.

smessmer replied:
It doesn't. This comment was meant to be 3 lines down at the one-element constructor. I fixed it.

smessmer replied:
About the `implicit` comments, explicit/implicit only matters for one-argument constructors. I could leave the `implicit` comment at the `OneElt` constructor, but that's actually the one with the TODO to make explicit, that's why I removed the comment.

ezyang commented on aten/src/ATen/ArrayRef.h:
```diff
@@ -23,170 +24,184 @@
+  constexpr ArrayRef(const T& OneElt) : Data(&OneElt), Length(1) {}
+
+  /// Construct an ArrayRef from a pointer and length.
+  constexpr ArrayRef(const T* data_, size_t length)
```
The renaming of the argument here seems both gratuitous, and wrong (aren't underscore suffixed identifiers supposed to be members?!)

smessmer replied:
fixed

ezyang commented on aten/src/ATen/ArrayRef.h:
```diff
@@ -23,170 +24,184 @@
+
+  /// front - Get the first element.
+  C10_CPP14_CONSTEXPR const T& front() const {
+    AT_CHECK(!empty(), "Empty list!");
```
The old error message seemed much more informative than the new one

smessmer replied:
fixed

ezyang commented on aten/src/ATen/ArrayRef.h:
```diff
@@ -23,170 +24,184 @@
+  /// slice(n, m) - Chop off the first N elements of the array, and keep M
+  /// elements in the array.
+  C10_CPP14_CONSTEXPR ArrayRef<T> slice(size_t N, size_t M) const {
+    AT_CHECK(N + M <= size(), "N = ", N, "; M = ", M, "; size = ", size());
```
Ditto here; why would you remove perfectly good explanatory text?

smessmer replied:
tbh, I have no idea why and how this was changed.

ezyang commented:
Code organization question: in this PR, you moved C++17.h from caffe2 folder to aten folder. But isn't this file supposed to be shared between both Caffe2 and PyTorch?

It's inappropriate to put C++17.h in ATen directly, because ATen is too heavy a dependency for us to put in mobile. It's best for us to have some directory which doesn't include anything outside of it, which is OK to ship to mobile (and we promise to keep an eye on binary sizes here). This directory should be depended on by both ATen and Caffe2, and in the near future Tensor and Storage classes will also live in this directory.

smessmer commented:
About code organization, caffe2 already has this weird pattern that they include .cpp files on a per-file basis to the mobile build. There is no clear separation in the folder structure for files that are or aren't used in mobile. I'm, however, introducing an ATen/utils folder and plan to put this file there. It's introduced in https://github.com/pytorch/pytorch/pull/9897 and I will move ArrayRef there.

I wasn't doing it in this PR, because I didn't figure out yet how to set the build system up for this.

ezyang commented:
Update for those paying attention: we're going to call the directory `ATen/core` now.

smessmer committed:
Improve ArrayRef (#9610)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/9610

Mostly making some stuff in ArrayRef constexpr to give it better perf.

Differential Revision: D8926785

fbshipit-source-id: 5403e86e4b1849718f24f0c79b6e70380b3eceaa
